### PR TITLE
Refactor event propagation

### PR DESCRIPTION
**Pull request by tindzk**

### CODE REVIEW

File: core/src/main/scala/org/scalarelational/Session.scala [hunk 1]
```diff
@@ -12,12 +12,13 @@ case class Session(datastore: Datastore, var inTransaction: Boolean = false) {
...
-  def hasConnection = _connection.nonEmpty
-  def connection = _connection match {
+  def hasConnection: Boolean = _connection.nonEmpty
+
+  def connection: Connection = _connection match {
     case _ if disposed => throw new RuntimeException("Session is disposed.")
     case Some(c) => c
     case None => {
-      val c = datastore.dataSource.getConnection
+      val c = datastore.dataSource.get.getConnection
```
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/Session.scala]: wouldn't this be better as: `val c = datastore.dataSource.map(_.getConnection)` so you don't then have to wrap it in an Option and then just call `c.get` at the end?

File: core/src/main/scala/org/scalarelational/extra/HikariSupport.scala [hunk 2]
```diff
@@ -8,27 +8,26 @@ import org.scalarelational.model.SQLDatastore
...
-      case ds: HikariDataSource => // Ignore HikariDataSource
-      case ds => {
-        val config = new HikariConfig()
-        config.setDataSource(ds)
-        val hikari = new HikariDataSource(config)
-        hikari.suspendPool()
-        dataSourceProperty := hikari
-      }
-    }
+  dataSourceProperty.filterCycles.attach {
```
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: `filterCycles` is a really confusing name for modifying the value on-change.

File: core/src/main/scala/org/scalarelational/extra/HikariSupport.scala [hunk 3]
```diff
@@ -8,27 +8,26 @@ import org.scalarelational.model.SQLDatastore
...
-  // Clean up data source after shutdown
-  override def dispose() = {
+  /** Clean up data source after shutdown */
+  override def dispose() {
     super.dispose()
 
     dataSource match {
-      case ds: HikariDataSource => ds.shutdown()
-      case _ => // Not a HikariDataSource
+      case Some(ds: HikariDataSource) => ds.shutdown()
```
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: Doesn't this lose type-safety?  Won't this always evaluate to true if `Some` is the type?

File: core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala [hunk 4]
```diff
@@ -1,16 +1,18 @@
...
+ * Persistent Properties allows key/value pairs to be persisted to a table for
+ * later retrieval and modification.
  *
- * Convenience methods are provided to get, set, and remove the value for a property and to create a Property instance
- * that can update the database in an event-driven manner.
+ * Convenience methods are provided to get, set, and remove the value for a
+ * property and to create a Property instance that can update the database in an
+ * event-driven manner.
  *
  * @author Matt Hicks <matt@outr.com>
```
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala]: You should really start adding your `@author` tag to code you add to.

File: core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala [hunk 5]
```diff
@@ -23,45 +25,45 @@ trait PersistentProperties extends Datastore {
...
 
     def apply(name: String): String = {
-      get(name).getOrElse(throw new NullPointerException(s"Unable to find $name in persistent properties table."))
+      get(name).getOrElse(
+        throw new NullPointerException(s"Unable to find $name in persistent properties table."))
     }
 
-    def update(name: String, newValue: String) {
-      val m = merge(
+    def update(name: String, newValue: String): Int =
```
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala]: Not relevant to this ticket, but this is going to have to change since not all databases support `Merge` apparently.

File: core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala [hunk 6]
```diff
@@ -23,45 +25,45 @@ trait PersistentProperties extends Datastore {
...
-      val p = Property[String](default = Some(get(key).getOrElse(default)))
-      p.change.on {
-        case evt => if (evt.newValue != null) {
-          this(key) = evt.newValue
-        } else {
-          remove(key)
-        }
+    def stringProperty(key: String): Opt[String] = {
+      val ch = new Opt[String](get(key))
+      ch.silentAttach {
```
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala]: Your naming really needs some work.

File: h2/src/main/scala/org/scalarelational/h2/H2Datastore.scala [hunk 7]
```diff
@@ -18,40 +16,31 @@ import org.scalarelational.table.Table
...
                      username: String = "sa",
-                     password: String = "sa") = {
+                     password: String = "sa") {
     this()
     dbUsername := username
     dbPassword := password
     modeProperty := mode
   }
 
-  protected def this(dataSource: DataSource) = {
```
Comment (darkfrog26) [inline h2/src/main/scala/org/scalarelational/h2/H2Datastore.scala]: Why did you remove this? It is a perfectly acceptable scenario to want to instantiate a `H2Datastore` and supply your own `DataSource`

File: h2/src/test/scala/org/scalarelational/h2/modular/ModularSpec.scala [hunk 8]
```diff
@@ -32,57 +32,57 @@ class ModularSpec extends WordSpec with Matchers {
       }
     }
     "connect handlers" in {
-      users.handlers.inserting.on {
+      users.handlers.inserting.attach {
```
Comment (darkfrog26) [inline h2/src/test/scala/org/scalarelational/h2/modular/ModularSpec.scala]: I think `on` is clearer what the purpose is than `attach` and is much more consistent with other event systems. Not a big deal really, but I presume you'd want the syntax as clear as possible.

File: project/ScalaRelationalBuild.scala [hunk 9]
```diff
@@ -5,7 +5,7 @@ object ScalaRelationalBuild extends Build {
   import Dependencies._
 
   lazy val root = Project(id = "root", base = file(".")).settings(name := "ScalaRelational", publish := {}).aggregate(core, h2, mysql, postgresql, mapper, versioning)
-  lazy val core = project("core").withDependencies(powerscala.property, hikariCP, scalaTest).settings(
+  lazy val core = project("core").withDependencies(powerscala.property, hikariCP, scalaTest, metaRx).settings(
```
Comment (darkfrog26) [inline project/ScalaRelationalBuild.scala]: Why is `powerscala.property` still in-place?

**Comment by darkfrog26** [comment]
This is a pretty massive change and though I said I was fine with abandoning Properties and Processors it will take a little bit of convincing to prove that this is a better alternative.

File: core/src/main/scala/org/scalarelational/Session.scala (hunk 1 above)
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/Session.scala]: Good catch.

File: core/src/main/scala/org/scalarelational/extra/HikariSupport.scala [hunk 10]
```diff
@@ -8,27 +8,26 @@ import org.scalarelational.model.SQLDatastore
...
-        dataSourceProperty := hikari
-      }
-    }
+  dataSourceProperty.filterCycles.attach {
+    case Some(ds) =>
+      val config = new HikariConfig()
+      config.setDataSource(ds)
+
+      val hikari = new HikariDataSource(config)
+      // hikari.suspendPool()  TODO Throws an exception
```
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: Do you think it is safe to comment out this line?

Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: I believe so...I'm honestly not sure why I put it in there.

File: core/src/main/scala/org/scalarelational/extra/HikariSupport.scala (hunk 2 above)
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: ``filterCycles`` does exactly what its name says. When we do ``dataSourceProperty := hikari`` later, the handler will not be called (because it would lead to cyclic behaviour). In your version you dealt with this by providing an explicit ``case`` block checking for ``HikariDataSource`` values.

Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: It still seems confusing to me...

File: core/src/main/scala/org/scalarelational/extra/HikariSupport.scala (hunk 3 above)
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: We don't lose type-safety. You're confusing this with generic parameters for which the compiler performs type erasure.

File: core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala (hunk 4 above)
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala]: I thought ``@author`` designates only the original author of the file.

File: core/src/main/scala/org/scalarelational/extra/HikariSupport.scala (hunk 3 above)
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: So this will not return true for `Some(AnotherDataSource)`?

File: core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala (hunk 6 above)
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala]: I don't like the naming either. Any ideas how to rename it? ``silentAttach`` does not call the handler for the initial value. It is equivalent to ``ch.tail.attach()``.

File: core/src/main/scala/org/scalarelational/model/ModularSupport.scala [hunk 11]
```diff
@@ -11,19 +12,28 @@ import org.scalarelational.instruction._
  *
  * @author Matt Hicks <matt@outr.com>
  */
-trait ModularSupport extends SQLContainer with Listenable with DataTypes {
+trait ModularSupport extends SQLContainer {
+  case class Processor[T]() {
```
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/model/ModularSupport.scala]: What do you think about ``Processor``? I would like to move it upstream.

File: core/src/main/scala/org/scalarelational/model/ModularSupport.scala [hunk 12]
```diff
@@ -11,19 +12,28 @@ import org.scalarelational.instruction._
...
+    val functions = ArrayBuffer.empty[T => T]
+    def fire(value: T): T =
```
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/model/ModularSupport.scala]: I would like to rename this method. In MetaRx I'm using ``produce()`` for channels.

File: h2/src/main/scala/org/scalarelational/h2/H2Datastore.scala (hunk 7 above)
Comment (tindzk) [inline h2/src/main/scala/org/scalarelational/h2/H2Datastore.scala]: I couldn't find any occurrences where this constructor was used. I thought it is redundant as ``dataSourceProperty`` is ``public``.

File: project/ScalaRelationalBuild.scala (hunk 9 above)
Comment (tindzk) [inline project/ScalaRelationalBuild.scala]: I forgot to delete it, thanks.

File: core/src/main/scala/org/scalarelational/extra/HikariSupport.scala (hunk 2 above)
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: Ok, how would you solve it instead?

File: core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala (hunk 6 above)
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala]: clarify what `handler` is in this context...

Comment (tindzk) [inline core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala]: The callback passed to ``silentAttach``.

File: h2/src/main/scala/org/scalarelational/h2/H2Datastore.scala (hunk 7 above)
Comment (darkfrog26) [inline h2/src/main/scala/org/scalarelational/h2/H2Datastore.scala]: It is moderately redundant, but I think I want to try to make `dataSource` immutable in the future so it will need to be supported. Also, just because it's not being used in ScalaRelational doesn't mean someone utilizing the framework wouldn't prefer to do it this way.

File: core/src/main/scala/org/scalarelational/model/ModularSupport.scala (hunk 11 above)
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/model/ModularSupport.scala]: Seems incredibly oversimplified. How is this better than `UnitProcessor` in powerscala? Can you prove that this is substantially faster?

File: core/src/main/scala/org/scalarelational/model/ModularSupport.scala (hunk 12 above)
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/model/ModularSupport.scala]: I think we should have a separate design discussion regarding Properties and Processors. So far I see this PR making drastic changes simply for the sake of removing powerscala while I see absolutely no advantage and several disadvantages.

**Comment by darkfrog26** [comment]
I think you're approaching this task in the wrong way. You seem to be entirely absorbed with replacing what you perceive is the current level of functionality powerscala when you should be offering a potential gain.  I see no gain but I see several losses of functionality and clarity.

File: core/src/main/scala/org/scalarelational/extra/HikariSupport.scala (hunk 3 above)
Comment (tindzk) [inline core/src/main/scala/org/scalarelational/extra/HikariSupport.scala]: Exactly. For example, this will return `Int`:

```scala
val y = Option[Any](42)
y match {
  case Some(x: String) => "String"
  case Some(x: Int)    => "Int"
  case _               => "Other"
}
```

File: core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala (hunk 4 above)
Comment (darkfrog26) [inline core/src/main/scala/org/scalarelational/extra/PersistentProperties.scala]: Since I made this note I have changed to no longer add @author tags at all.

**Comment by darkfrog26** [comment]
Before merging this I think it's worthwhile to run the performance tests before and after to make sure that we're gaining performance and not losing it. If nothing else, it would be useful to see how much this benefits us.
